### DISCUSSION TITLE
Crazyflie using kalman filter suddenly stopped while recieving commands

### DISCUSSION THREAD

**Discussion by cc299792458** [bitcraze/discussions#624] [Q&A] [answered]
Hi folks,

I encountered a strange thing that the crazyflie suddenly stopped while recieving commands.(like shown in the video)

I use kalman-filter as state estimeter and pid controller as controller.
I use bigquad deck to drive the motors.
The performance of the crazyflie used to be right. But after I modifed some values in the firmware "kalman_supurvisor.c", the strange thing happened.

Please see picture I attached. Because the kalman filter will reset itself during the time I tune PID gain, I try to increase the bound value or just simply "return true" to prevent kalman filter reset itself. So I guess the reason it stopped may relate to this.

I continuesly send "roll = 0, pitch = 0, yawrate = 0, thrust = 10000" to crazyflie.(I know there is an api can do the same thing but actually this command is writen by myself(both python script and firmware), I'm sure it can work well.)

Could you please give me some advises? 

Thanks a lot!
cc299792458

https://user-images.githubusercontent.com/52767591/226297247-0dbfe9c9-4a47-4c00-b85e-30cc56f81ad3.MOV

![Screenshot from 2023-03-20 17-26-06](https://user-images.githubusercontent.com/52767591/226299771-35db3e08-d791-4d45-850d-97a881a56bef.png)

**Comment by cc299792458** [comment]
I find the video can not play on the website, maybe you can download it. Thanks!

**Comment by krichardsson** [comment, marked as answer]
Just to clarify, when the motors stop in your video, is this when you get a reset of the kalman estimator?

Your solution of returning `true` in `kalmanSupervisorIsStateWithinBounds()` should work just fine. Are you sure this is what causes the problems? Could it be something else instead?

I assume you do not have any position input (lighthouse, Loco or similar)? Without position input, the estimated position in the estimator tends to drift off, and when reaching 100m, the supervisor thinks it is unreasonable and resets the estimator.
 
Could you use the complementary estimator instead of the kalman estimator while tuning the PID?

Have you tried to tried to remove functionality to reduce any complexity (or use the stock firmware/lib) to see if you still see the same problem?

**Reply by cc299792458** [reply]
Hi, thanks for your kindly reply!

For the first question: I printed console information so I'm sure when the motors stopped, the kalman estimator didn't reset.(Because I increase the bounds. On the other hand, this phenomenon also happened when I just "return true" in which case the kalman estimator would never reset).

For the second question: I tried to switch to complementary estimator and this phenomenon disappeared so I'm more convinced that this problem was caused by kalman estimator.

For the third question: Although I didn't have a localization system during tuning attitude and attitude_rate PID params, I will use Vicon in the next steps when tuning the params of position PID and velocity PID. When I tuned the attitude and attitude_rate PID params, I continuously sent x=0, y=0, z=0 to crazyflie.

And for the last question, I considered switch to complementary estimator when tuning the PID but I was afraid this would affect the downstream tasks because I have to use kalman estimator back to use Vicon data.

**Reply by krichardsson** [reply]
Thanks for the information!
I do not know what might be causing this, but let's try to figure it out :-)

> I considered switch to complementary estimator when tuning the PID but I was afraid this would affect the downstream tasks because I have to use kalman estimator back to use Vicon data.

I see, let's stay with the kalman estimator and try to figure out what is happening.


I think a good first step would be to log the thrust that the controller is requesting, this would help us understand if the stopping of motors is related to the estimator/controller or if it happens after the controller. My guess is that it is happening somewhere in the estimator/controller, but this measurement would help us to rule out the code executing after the controller.

You can log the thrust output of the controller using the `stabilizer.thrust` log variable.
Also log the `motor.m1req` (and `m2req` - `m4req` if you want to) to see what we request from the motors.

**Reply by cc299792458** [reply]
Thanks for your advice!
I tried logging "stablizer.thrust" and plotted it, as you can see in the pitcutre I attached.

Let me explain the pitcure.
The experiment lasted for 10 seconds. During the first 5 seconds, I sent command roll=0, pitch=0, yawrate=0 and thrust=5000 while during the 5 to 10 seconds, I sent command roll=0, pitch=5, yawrate=0 and thrust=5000. 
The unit of x-axis is millisecond. 
In subplot(3, 3), I log the data of thrust(stabilizer.thrust). As you can see, there are 2 place that the value of it is zero. That's exact when the motors stopped.
In subplot(1, 2), I log the data of pitch(stateEstimate.pitch) and pitch_reference(controller.pitch). 
In subplot(2, 2), I log the data of pitchrate(controller.r_pitch, I translate it from radian to degree) and pitchrate_reference(controller.pitchrate).
In subplot(3, 1) and (3, 2), I log the data of motor(motor.m1, motor.m2, motor.m3, motor.m4).
Other subplots are useless in this plot because I didn't log those variables, just plotted 0.

It seems like that something "make the whole platform stopped"(?), because the thrust turns to zero although my command of thrust is 5000.
Could you please share your opinion with me?
  
![Screenshot from 2023-03-22 20-11-33](https://user-images.githubusercontent.com/52767591/226947112-1f3449ab-b7f2-430f-90c6-a9afeeb13ead.png)

**Reply by krichardsson** [reply]
Great, thanks for the measurements!
As you mentioned, we can see that the thrust output from the controller drops and I think we can assume the problem is related to the estimator/controller. I also think that your log of stateEstimate.pitch tells us that the estimator is OK and that the problem is related to the controller.

There are not that many places the thrust is modified in the pid controller (where it can be set to zero), you can find them in [`controller_pid.c`](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/controller_pid.c) and [position_controller_pid.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/position_controller_pid.c) if you want to check them out. 

There is another possible explanation as well. There is a safety mechanism that is intended to stop the motors if the crazyflie loses radio connection. It is a simple timeout that checks that setpoints are continuously received, and if the time between setpoints is too long, the motors will be stopped. I don't remember exactly what this time out is, but if you send a setpoint every 0.2 seconds (5 Hz) that should be enough.

**Reply by cc299792458** [reply]
Thanks for the instruction, sir!

I checked controller_pid.c and position_controller_pid.c and I think it's a little strange why thrust can be set to zero.
'stabilizer.thrust' we logged equals to 'control->thrust' in controller_pid.c. The only codes to modified 'control->thrust' in controller_pid.c are line 142 and line 146 in my pitcure 1.
So we can infer that the only way to set control->thrust to zero is set actuatorThrust to zero.
![Screenshot from 2023-03-24 14-24-58](https://user-images.githubusercontent.com/52767591/227442223-ca700bce-b5e0-42c3-8c4e-8a9002f1aa22.png)

There are also only 2 ways to modify the value of actuatorThrust in controller_pid.c, which is line 93 and line 99 in my pitcure 2. The first one is modified by positionController but it will be directly overwrote by line setpoint->thrust in line 99 because I continously sent thrust=5000 to it. 
![Screenshot from 2023-03-24 14-30-24](https://user-images.githubusercontent.com/52767591/227443034-5f473347-2dc4-4285-a405-ff3c7e26f62c.png)

So how can control->thrust be set to 0? 
Another clue is, not only the control->thrust was set to 0, but also the control->roll, control->pitch, and control->yaw were set to 0. So if the problem really lies with the controller, I think it's more possible to do with line 144(in pitcture 1).

You point out that the estimator seems correct. But we can see that when the thrust drops to zero, the noise of pitch data decrease obviously, is that correct? So I still suspect that there is a problem with the state estimator.

I have another guess. From my video, you can see that the dynamic parameters of my quadrotor are significantly different from ordinary crazyflie. Maybe there are significant errors in the calculation of the Kalman filter due to problems with dynamic parameters?

For the last thing, I know what you said that crazyflie will stop after 2 seconds without receiving new setpoint. But I think the frequency I sent the setpoint is higher enough.

That's all clues. I konw it maybe a little complicated. Could you just give me some more suggestions? I am very grateful for your kindness and help!

**Reply by krichardsson** [reply]
> For the last thing, I know what you said that crazyflie will stop after 2 seconds without receiving new setpoint. But I think the frequency I sent the setpoint is higher enough.

Before digging too deep into the controller and estimator code, I would suggest to make sure the timeout is not triggered. It is easy to test and would be nice to rule out. 

The code for the timeout can be found in the [`commander.c`](https://github.com/bitcraze/crazyflie-firmware/blob/1346781770e7ccfa90e702b061a44f89de39353d/src/modules/src/commander.c#L107-L121). After 0.5 seconds, the CF should level out and after 2 seconds the motors should shut down. In the stabilizer loop we call [`commanderGetSetpoint()`](https://github.com/bitcraze/crazyflie-firmware/blob/1346781770e7ccfa90e702b061a44f89de39353d/src/modules/src/stabilizer.c#L291) and if the time out has occured the setpoint will be cleared
The best way to test this is probably to log the setpoint data and see if the thrust is set to 0, if it is I think the timeout is the problem. There is a [log group for the setpoint in the stabilizer loop](https://github.com/bitcraze/crazyflie-firmware/blob/1346781770e7ccfa90e702b061a44f89de39353d/src/modules/src/stabilizer.c#L383-L444), but thrust is unfortunately not part of it. You can add it by simply adding the line `LOG_ADD_CORE(LOG_FLOAT, thrust, &setpoint.thrust)` in the group and log `ctrltarget.thrust`

**Reply by cc299792458** [reply]
Long time no contact, sir! I'm sorry for that.

The reason why I stop replying this discussion is, the "strange phenomenon"(suddenly shut down) disappeared itself. 

The whole process is:
One night, I uncomment the code "return true" in kalmanSupervisorIsStateWithinBounds() which makes it always return true.
Then, just as what I said, "suddenly shut down phenomenon" never happened again.
Actually, this is also strange I think, because I do remember in 1 version of my trials, I have tried to return true forever but the phenomenon happened at that time.

But you know, we can't debug for an unrepeatable thing.

However, I still have some guesses about the "suddenly shut down phenomenon", though I'm not pretty sure. Since I indeed didn't change a lot of code of firmware, I think it's maybe a problem related to either communication(between PC and Crazyflie) or hardware. The reason is, though the "suddenly shut down phenomenon" never happens, sometimes the communication will stop automatically. (I inferred this because the blinking red led stop shining and the reference argument logged from Crazyflie turned to zeros). Another possible reason is the hardware, but it needs some further trial like another board.  

I'm not pretty sure, just for a guess.

Thanks a lot for your kindly help!!
